app/services/submission/publisher.rb

```diff
@@ -0,0 +1,46 @@
+class Submission::Publisher
+  attr_accessor :submission
+
+  def initialize(submission)
+    raise ArgumentError, "Submission not finalized" unless submission.finalized?
+
+    self.submission = submission
+  end
+
+  def publish
+    raise ArgumentError, "Submission is published" if submission.publishable?
+    now = Time.zone.now
+    submission.transaction do
+      submission.update_attributes!(publishable: true)
+      publish_object(submission.submitted_object, now)
+      associated_collections.each do |collection|
+        collection.not_published.each { |instance| publish_object(instance, now) }
```

Comment:
There might be thousands of trait scores published with a submission. Having them published one by one may take some time. Would it be possible to optimize it a little with with update_all?